sympy/sets/sets.py

```diff
@@ -1431,8 +1434,14 @@ def reduce(args):
         # all other sets in the intersection
         for s in args:
             if s.is_FiniteSet:
-                return s.func(*[x for x in s
-                    if all(other.contains(x) == True for other in args)])
+                args = [a for a in args if a != s]
+                res = s.func(*[x for x in s
```

Comment:
`res = FiniteSet(*[...])` will be more readable, it also goes with the Zen of Python [#1] `Explicit is better than Implicit`

[#1]: https://www.python.org/dev/peps/pep-0020/